### DISCUSSION TITLE
Feature: Add `FileSuggester` for `file` in `ToolbarSettingsModal` for better file referencing

### DISCUSSION THREAD

**Discussion by LBF38** [chrisgurney/obsidian-note-toolbar#26] [Ideas]
Hi,
I'm proposing an enhancement of the `ToolbarSettingsModal` to use a `FileSuggester` rather than a plain text field.

https://github.com/chrisgurney/obsidian-note-toolbar/blob/6ac92be6920215cb8d61608adeb0783f00ab3ab7/src/Settings/ToolbarSettingsModal.ts#L272-L276

This will allow the user to better search his vault files, hence being sure of what file he references.

I might create a PR in the next few days if I've got time and if you agree with this feature.

Thanks for considering it !

(btw, this plugin is a great idea ! Thanks a lot !)

**Comment by chrisgurney** [comment]
@LBF38 Thanks for the idea!

So I originally had written a class to do this -- inspired by [code from Templater](https://github.com/SilentVoid13/Templater/blob/master/src/settings/suggesters/FileSuggester.ts) -- but then read [this guideline](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+iterating+all+files+to+find+a+file+by+its+path) and abandoned the idea.

If there is an updated, official way to do this, or something more efficient, I'd love to do this! (My search of the API docs and Discord are coming up short.)

I've [reached out on Discord](https://discord.com/channels/686053708261228577/840286264964022302/1227232035081424917) so we'll see what comes back, unless you've got any ideas. 😄

**Comment by chrisgurney** [comment]
@LBF38 It _sounds_ like I may have misinterpreted the guidelines, but I'm still waiting to see if I'll get official confirmation from the team.

I've drafted a commit that will be part of the next beta, here: https://github.com/chrisgurney/obsidian-note-toolbar/commit/ae407a00a365584c6a0a862fb28ca6d506226313

...but I'm not sure how performant this will be on a large vault. I will need some help testing that.

I do appreciate the offer of writing a PR! Thanks again!

**Reply by LBF38** [reply]
Hi @chrisgurney , 
I think that the code you drafted makes perfect sense and is aligned with the plugin's guidelines.
You are using Obsidian's APIs so it should be working seamlessly while adhering to what the team wants/recommends.
I agree on having the official confirmation.

Another approach on that would be to first prompt the user for the folder in which the file is located and then prompt for the specific file I think ? 
I don't see any other solution.

For the performance testing, we can generate a large vault and see how much time it takes to respond on random files.
What do you think ?

**Comment by chrisgurney** [comment]
The file suggester is now available [in the beta](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/40).

@LBF38 thank you for the suggestion for testing a large vault. If you have any free time I'd appreciate it if you could give that a spin, otherwise no worries!

**Reply by LBF38** [reply]
I've tested with +5k files in a vault and it works perfectly ! (vault generated for testing it)
It is really quick and we can easily search any file.
I think that performance comes from the cached index of all files in the vault by Obsidian.

Thanks for implementing it ! 👏

**Reply by chrisgurney** [reply]
Really appreciate you doing that. I can sleep easy tonight. 😆 Thank you!

**Comment by chrisgurney** [comment]
This feature has now been [released](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/40).
